boards: sensry: Add ganymed_bob board support

tswaehn opened the pull request:
Add soc support for Sensry's RISCV32 based SY1xx.
Variants of the soc are GBM and GEN1.
Add board support for eval board ganymed_bob, which is a break-out-board for both soc variants.

fixes: #77757

nordicjm commented on boards/sensry/ganymed_bob/board.yml:
```diff
@@ -0,0 +1,9 @@
+  name: ganymed_bob
+  vendor: sensry
+  socs:
+    - name: sy120_gbm
+    - name: sy120_gen1
```
why isn't this a variant if all it does is add sensors?

tswaehn replied:
as the sensors are placed inside the chip, we can use the chip on different boards. so the sensors come with the chip and not with the board.

nordicjm replied:
2 SoCs it is then

nordicjm commented on boards/sensry/ganymed_bob/doc/index.rst:
```diff
@@ -0,0 +1,141 @@
+Overview
+********
+
+.. note::
+
+   All software for the nGanymed BOB is experimental and hardware availability
+   is restricted to the participants in the limited sampling program.
```
this looks like you've taken the document from a nordic board, unless this is true, it should be removed

tswaehn replied:
yes, its taken from nordic template. its by intention. we start to introduce the chip.

nordicjm commented on boards/sensry/ganymed_bob/doc/index.rst:
```diff
@@ -0,0 +1,141 @@
+Supported Features
+==================
+
+The variant ``ganymed-bob/sy120-gbm/cpuapp`` board configuration supports the following
```
*board target, the board target is also actually wrong

tswaehn replied:
fixed

nordicjm commented on boards/sensry/ganymed_bob/doc/index.rst:
```diff
@@ -0,0 +1,141 @@
+
+Other hardware features have not been enabled yet for this board.
+
+The variant ``ganymed-bob/sy120-gen1/cpuapp`` board configuration supports additionally to the ``ganymed-bob/sy120-gbm/cpuapp`` the following
```
as above

tswaehn replied:
fixed

nordicjm commented on boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.dts:
```diff
@@ -0,0 +1,20 @@
+
+/ {
+
+
```
excess spacing

tswaehn replied:
fixed

nordicjm commented on boards/sensry/ganymed_bob/ganymed_bob_sy120_gbm.yaml:
```diff
@@ -0,0 +1,16 @@
+arch: riscv
+toolchain:
+  - zephyr
+
```
```suggestion
```

tswaehn replied:
fixed

nordicjm commented on boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.dts:
```diff
@@ -0,0 +1,20 @@
+
+/ {
+
+
```
as above

tswaehn replied:
fixed

nordicjm commented on boards/sensry/ganymed_bob/ganymed_bob_sy120_gen1.yaml:
```diff
@@ -0,0 +1,16 @@
+arch: riscv
+toolchain:
+  - zephyr
+
```
as above

nordicjm commented on drivers/serial/Kconfig.sy1xx:
```diff
@@ -0,0 +1,10 @@
+config UART_SY1XX
+	bool "Sensry SY1xx driver"
+	default y
+    depends on DT_HAS_SENSRY_SY1XX_UART_ENABLED
```
tabs in Kconfig

tswaehn replied:
fixed

nordicjm commented on soc/sensry/ganymed/Kconfig:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2024 sensry.io
+
+config SOC_FAMILY_GANYMED
+	select BUILD_OUTPUT_HEX
+	select BUILD_OUTPUT_BIN
```
why are both of these always required?

tswaehn replied:
actually really need only binary. removed the hex options

nordicjm commented on soc/sensry/ganymed/Kconfig.soc:
```diff
@@ -0,0 +1,12 @@
+
+config SOC_SERIES
+	default "sy1xx" if SOC_SERIES_SY1XX
+
```
1x empty line not 2

tswaehn replied:
fixed

nordicjm commented on soc/sensry/ganymed/sy1xx/Kconfig:
```diff
@@ -0,0 +1,147 @@
+config SOC_SERIES_SY1XX
+    bool
+	select RISCV
+    select RISCV_ISA_RV32I
```
wrong indent

tswaehn replied:
fixed

nordicjm commented on soc/sensry/ganymed/sy1xx/Kconfig:
```diff
@@ -0,0 +1,147 @@
+
+if SOC_SERIES_SY1XX
+
+config RISCV_PMP
```
these go in the `Kconfig.defconfig` file

tswaehn replied:
done

nordicjm commented on soc/sensry/ganymed/sy1xx/Kconfig:
```diff
@@ -0,0 +1,147 @@
+    default 10
+
+
+## general
+config BOOT_BANNER
+    default y
```
what?

tswaehn replied:
removed

nordicjm commented on soc/sensry/ganymed/sy1xx/Kconfig:
```diff
@@ -0,0 +1,147 @@
+
+## serial
+config SERIAL_SUPPORT_INTERRUPT
+    default n
+
+config UART_ASYNC_API
+    default n
+
+config SERIAL_HAS_DRIVER
+    default y
+
+config SERIAL
+    default y
+
+## console
+config CONSOLE
+    default y
+
+config UART_CONSOLE
+    default y
+
+config CONSOLE_HANDLER
+    default y
+
+config PRINTK
+    default y
+
+config STDOUT_CONSOLE
+    default y
+
+config EARLY_CONSOLE
+    default y
+
+
+## shell
+config SHELL
+    default y
+config KERNEL_SHELL
+    default y
+config SHELL_BACKEND_SERIAL
```
No to any of this

tswaehn replied:
moved this to the board, should not be in soc

nordicjm commented on soc/sensry/ganymed/sy1xx/common/event-unit.h:
```diff
@@ -0,0 +1,8 @@
+
+#ifndef MY_ZEPHYR_APP_LOCAL_ZEPHYRPROJECT_LATEST_ZEPHYR_SOC_SENSRY_GANYMED_SY1XX_COMMON_EVENT_UNIT_H
+#define MY_ZEPHYR_APP_LOCAL_ZEPHYRPROJECT_LATEST_ZEPHYR_SOC_SENSRY_GANYMED_SY1XX_COMMON_EVENT_UNIT_H
+
```
file can go

tswaehn replied:
removed

tswaehn commented on boards/sensry/ganymed_bob/Kconfig.defconfig:
```diff
@@ -0,0 +1,2 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2024 sensry.io
```
done

nordicjm replied:
??

tswaehn replied:
missed that again. sure, can go.

tswaehn commented:
@nordicjm I think formal, code style related things should be covered now. its actually great you spent the time for a feedback. thank you!

dcpleung commented on boards/sensry/ganymed_bob/Kconfig.defconfig:
```diff
@@ -0,0 +1,65 @@
+	default 10
+
+## serial
+config SERIAL_SUPPORT_INTERRUPT
+	default n
```
This should be "selected" by driver and no need to explicitly disable it.

dcpleung commented on boards/sensry/ganymed_bob/Kconfig.defconfig:
```diff
@@ -0,0 +1,65 @@
+## serial
+config SERIAL_SUPPORT_INTERRUPT
+	default n
+config UART_ASYNC_API
+	default n
```
`CONFIG_UART_ASYNC_API` is disabled by default so to explicitly do so.

tswaehn replied:
fixed

dcpleung commented on boards/sensry/ganymed_bob/Kconfig.defconfig:
```diff
@@ -0,0 +1,65 @@
+	default n
+config UART_ASYNC_API
+	default n
+config SERIAL_HAS_DRIVER
+	default y
```
This selected by driver and no need to explicitly enable it.

tswaehn replied:
understand

dcpleung commented on drivers/serial/uart_sy1xx.c:
```diff
@@ -0,0 +1,317 @@
+	return 0;
+}
+
+/*
+ * \return: < 0 ... error
+ *	    = 0 ok
+ *	    > 0 ... busy
+ */
```
I recommend changing this into proper doxygen style comments so IDE can display it during code completion, especially the return values.

ditto for those below.

tswaehn replied:
done

dcpleung reviewed (commented):
Could you split the commits into small ones? For example, one for timer driver, one for serial driver, one for SoC, one for board and one for dts? That would make it easier to review.

tswaehn committed:
soc: sensry: Add support for SY120-GBM and SY120-GEN1

Add soc support for Sensry's RISCV32 based SY1xx.
Variants of the soc are GBM and GEN1.

Signed-off-by: Sven Ginka <s.ginka@sensry.de>

tswaehn committed:
driver: serial: Add support for sy1xx

Add uart driver for Sensry's RISCV32 based SY1xx.

Signed-off-by: Sven Ginka <s.ginka@sensry.de>

tswaehn committed:
driver: timer: Add support for sy1xx

Add sys timer driver for Sensry's RISCV32 based SY1xx.

Signed-off-by: Sven Ginka <s.ginka@sensry.de>

tswaehn committed:
board: sensry: Add support for sy1xx

Add board support for eval board ganymed_bob, which
is a break-out-board for both soc variants.
Variants of the soc are GBM and GEN1.

Signed-off-by: Sven Ginka <s.ginka@sensry.de>

tswaehn commented:
dear @nordicjm, dear @dcpleung 

I followed you advice. Would be great to have your review of my changes.

nordicjm commented on dts/bindings/cpu/sensry,ganymed-sy1xx.yaml:
```diff
@@ -0,0 +1,6 @@
+
+description: Sensry Ganymed SY1xx Core CPU
+
+compatible: "sensry,sy1xx"
```
missing cpu.yaml include, probably missing riscv,cpus.yaml include too

tswaehn replied:
riscv,cpus.yaml includes cpu.yaml; included riscv,cpus.yaml

nordicjm commented on soc/sensry/ganymed/sy1xx/Kconfig:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2024 sensry.io
+
+config SOC_SERIES_SY1XX
+	bool
```
```suggestion
```

nordicjm commented on boards/sensry/ganymed_bob/doc/img/ganymed_bob_sy120_gbm.webp:
Put webp files through https://tinypng.com/ to reduce file size

tswaehn replied:
actually I am impressed. I was using GIMP before - which is THE image tool. and finally tinypng shrinked them by 45% without visual impact.

tswaehn commented:
I am really keen in getting the PR accepted. Just wondering if there is things open that I need to care about?

tswaehn commented:
I think to have covered all points from review. Is there anything I could do?

nordicjm reviewed (approved):
Board/SoC files OK

dcpleung commented on soc/sensry/ganymed/sy1xx/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+
+zephyr_include_directories(common)
+
+zephyr_compile_options(-march=rv32imc_zicsr -mabi=ilp32)
```
`cmake/compiler/gcc/target_riscv.cmake` would generate the correct `march` argument if you set `CONFIG_RISCV_ISA_EXT_*` correctly. So there is no need to specify `march` here.

tswaehn replied:
@dcpleung just added PR here: https://github.com/zephyrproject-rtos/zephyr/pull/78633

tswaehn replied:
this is merged, would it be ok to "Resolve Conversation" ?

dcpleung commented:
Another suggestion is to prefix the SoC macro like `UDMA_*` and `PAD_CONFIG` with the SoC name (or other more unique identifier). This would help others when searching for code as they can exclude such substrings.

fkokosinski commented:
Question: what about the maintainership status of this board? If I understand correctly, right now it's not covered by any of the preexisting areas of maintenance.

tswaehn commented:
> Question: what about the maintainership status of this board? If I understand correctly, right now it's not covered by any of the preexisting areas of maintenance.

I can do, ... just need to understand my duties.
